File: python/handler.py
import json
import requests
from pwrpy.pwrsdk import PWRPY
from database_service import DatabaseService

VIDA_ID = 73_746_238
RPC_URL = "https://pwrrpc.pwrlabs.io/"
REQUEST_TIMEOUT = 10

pwrpy_client = None
subscription = None
peers_to_check_root_hash_with = []

# Fetches the root hash from a peer node for the specified block number
def fetch_peer_root_hash(peer, block_number):
    url = f"http://{peer}/rootHash?blockNumber={block_number}"
        
    try:
        response = requests.get(url, timeout=REQUEST_TIMEOUT, headers={'Accept': 'text/plain'})
        
        if response.status_code == 200:
            hex_string = response.text.strip()
            
            if not hex_string:
                print(f"Peer {peer} returned empty root hash for block {block_number}")
                return False, None
            
            try:
                root_hash = bytes.fromhex(hex_string)
                print(f"Successfully fetched root hash from peer {peer} for block {block_number}")
                return True, root_hash
            except ValueError:
                print(f"Invalid hex response from peer {peer} for block {block_number}")
                return False, None
        else:
            print(f"Peer {peer} returned HTTP {response.status_code} for block {block_number}")
            return True, None
            
    except Exception:
        print(f"Failed to fetch root hash from peer {peer} for block {block_number}")
        return False, None

# Validates the local Merkle root against peers and persists it if a quorum of peers agree
def check_root_hash_validity_and_save(block_number):
    local_root = DatabaseService.get_root_hash()
    
    if not local_root:
        print(f"No local root hash available for block {block_number}")
        return
    
    peers_count = len(peers_to_check_root_hash_with)
    quorum = (peers_count * 2) // 3 + 1
    matches = 0
    
    for peer in peers_to_check_root_hash_with:
        success, root_hash = fetch_peer_root_hash(peer, block_number)
        
        if success and root_hash:
            if root_hash == local_root:
                matches += 1
        else:
            if peers_count > 0:
                peers_count -= 1
                quorum = (peers_count * 2) // 3 + 1
        
        if matches >= quorum:
            DatabaseService.set_block_root_hash(block_number, local_root)
            print(f"Root hash validated and saved for block {block_number}")
            return
    
    print(f"Root hash mismatch: only {matches}/{len(peers_to_check_root_hash_with)} peers agreed")
    DatabaseService.revert_unsaved_changes()
    subscription.set_latest_checked_block(DatabaseService.get_last_checked_block())

# Executes a token transfer described by the given JSON payload
def handle_transfer(json_data, sender_hex):
    try:
        amount = int(json_data.get('amount', 0))
        receiver_hex = json_data.get('receiver', '')
        
        if amount <= 0 or not receiver_hex:
            print(f"Invalid transfer data: {json_data}")
            return

        sender_address = sender_hex[2:] if sender_hex.startswith('0x') else sender_hex
        receiver_address = receiver_hex[2:] if receiver_hex.startswith('0x') else receiver_hex

        sender = bytes.fromhex(sender_address)
        receiver = bytes.fromhex(receiver_address)
        
        success = DatabaseService.transfer(sender, receiver, amount)
        
        if success:
            print(f"Transfer succeeded: {amount} from {sender_hex} to {receiver_hex}")
        else:
            print(f"Transfer failed (insufficient funds): {amount} from {sender_hex} to {receiver_hex}")
            
    except Exception as e:
        print(f"Error handling transfer: {e}")

# Processes a single VIDA transaction
def process_transaction(txn):
    try:
        data_hex = txn.data
        data_bytes = bytes.fromhex(data_hex)
        
        data_str = data_bytes.decode('utf-8')
        json_data = json.loads(data_str)
        
        action = json_data.get('action', '')
        
        if action.lower() == 'transfer':
            handle_transfer(json_data, txn.sender)
            
    except Exception as e:
        print(f"Error processing transaction: {e}")

# Callback invoked as blocks are processed
def on_chain_progress(block_number):
    DatabaseService.set_last_checked_block(block_number)
    check_root_hash_validity_and_save(block_number)
    print(f"Checkpoint updated to block {block_number}")
    DatabaseService.flush()

# Subscribes to VIDA transactions starting from the given block
def subscribe_and_sync(from_block):
    global pwrpy_client, subscription
    
    print(f"Starting VIDA transaction subscription from block {from_block}")
    
    pwrpy_client = PWRPY(RPC_URL)
    
    subscription = pwrpy_client.subscribe_to_vida_transactions(
        VIDA_ID,
        from_block,
        process_transaction,
        on_chain_progress
    )
    
    print(f"Successfully subscribed to VIDA {VIDA_ID} transactions")

File: javascript/handler.js
import PWRJS from "@pwrjs/core";
import fetch from 'node-fetch';
import DatabaseService from './databaseService.js';

const VIDA_ID = 73746238;
const RPC_URL = "https://pwrrpc.pwrlabs.io/";
const REQUEST_TIMEOUT = 10000;

export let peersToCheckRootHashWith = [];
let pwrjsClient = null;
let subscription = null;

// Fetches the root hash from a peer node for the specified block number
async function fetchPeerRootHash(peer, blockNumber) {
    const url = `http://${peer}/rootHash?blockNumber=${blockNumber}`;
    
    try {
        const response = await fetch(url, {
            method: 'GET',
            timeout: REQUEST_TIMEOUT,
            headers: {
                'Accept': 'text/plain'
            }
        });
        
        if (response.ok) {
            const hexString = await response.text();
            const trimmed = hexString.trim();
            
            if (!trimmed) {
                console.log(`Peer ${peer} returned empty root hash for block ${blockNumber}`);
                return { success: false, rootHash: null };
            }
            
            try {
                const rootHash = Buffer.from(trimmed, 'hex');
                console.log(`Successfully fetched root hash from peer ${peer} for block ${blockNumber}`);
                return { success: true, rootHash };
            } catch (error) {
                console.log(`Invalid hex response from peer ${peer} for block ${blockNumber}`);
                return { success: false, rootHash: null };
            }
        } else {
            console.log(`Peer ${peer} returned HTTP ${response.status} for block ${blockNumber}`);
            return { success: true, rootHash: null };
        }
    } catch (error) {
        console.log(`Failed to fetch root hash from peer ${peer} for block ${blockNumber}`);
        return { success: false, rootHash: null };
    }
}

// Validates the local Merkle root against peers and persists it if a quorum of peers agree
async function checkRootHashValidityAndSave(blockNumber) {
    const localRoot = await DatabaseService.getRootHash();
    
    if (!localRoot) {
        console.log(`No local root hash available for block ${blockNumber}`);
        return;
    }
    
    let peersCount = peersToCheckRootHashWith.length;
    let quorum = Math.floor((peersCount * 2) / 3) + 1;
    let matches = 0;
    
    for (const peer of peersToCheckRootHashWith) {
        const { success, rootHash } = await fetchPeerRootHash(peer, blockNumber);
        
        if (success && rootHash) {
            if (localRoot.equals(rootHash)) {
                matches++;
            }
        } else {
            if (peersCount > 0) {
                peersCount--;
                quorum = Math.floor((peersCount * 2) / 3) + 1;
            }
        }
        
        if (matches >= quorum) {
            await DatabaseService.setBlockRootHash(blockNumber, localRoot);
            console.log(`Root hash validated and saved for block ${blockNumber}`);
            return;
        }
    }
    
    console.log(`Root hash mismatch: only ${matches}/${peersToCheckRootHashWith.length} peers agreed`);
    await DatabaseService.revertUnsavedChanges();
    subscription.setLatestCheckedBlock(BigInt(await DatabaseService.getLastCheckedBlock()));
}

// Executes a token transfer described by the given JSON payload
async function handleTransfer(jsonData, senderHex) {
    const amount = BigInt(jsonData.amount || 0);
    const receiverHex = jsonData.receiver || "";
    
    if (amount <= 0 || !receiverHex) {
        console.log("Skipping invalid transfer:", jsonData);
        return;
    }

    const senderAddress = senderHex.startsWith("0x") ? senderHex.slice(2) : senderHex;
    const receiverAddress = receiverHex.startsWith("0x") ? receiverHex.slice(2) : receiverHex;
    
    const sender = Buffer.from(senderAddress, 'hex');
    const receiver = Buffer.from(receiverAddress, 'hex');
    
    const success = await DatabaseService.transfer(sender, receiver, amount);
    
    if (success) {
        console.log(`Transfer succeeded: ${amount} from ${senderHex} to ${receiverHex}`);
    } else {
        console.log(`Transfer failed (insufficient funds): ${amount} from ${senderHex} to ${receiverHex}`);
    }
}

// Processes a single VIDA transaction
function processTransaction(txn) {    
    try {
        const dataBytes = Buffer.from(txn.data, 'hex');
    
        const dataStr = dataBytes.toString('utf8');
        const jsonData = JSON.parse(dataStr);
        
        const action = jsonData.action || "";
        
        if (action.toLowerCase() === "transfer") {
            handleTransfer(jsonData, txn.sender);
        }
    } catch (error) {
        console.error("Error processing transaction:", txn.hash, error);
    }
}

// Callback invoked as blocks are processed
async function onChainProgress(blockNumber) {
    try {
        await DatabaseService.setLastCheckedBlock(blockNumber);
        await checkRootHashValidityAndSave(blockNumber);
        console.log(`Checkpoint updated to block ${blockNumber}`);
        await DatabaseService.flush();
    } catch (error) {
        console.error("Failed to update last checked block:", blockNumber, error);
    } finally {
        return null;
    }
}

// Subscribes to VIDA transactions starting from the given block
export async function subscribeAndSync(fromBlock) {
    console.log(`Starting VIDA transaction subscription from block ${fromBlock}`);

    // Initialize RPC client
    pwrjsClient = new PWRJS(RPC_URL);
    
    // Subscribe to VIDA transactions
    subscription = pwrjsClient.subscribeToVidaTransactions(
        VIDA_ID,
        BigInt(fromBlock),
        processTransaction,
        onChainProgress
    );
    console.log(`Successfully subscribed to VIDA ${VIDA_ID} transactions`);
}
